src/RevShareContractsUpgrader.sol

```diff
@@ -0,0 +1,113 @@
+/// @dev    Supports two operations:
+///         1. setupRevShare() - Setup revenue sharing on already-upgraded contracts
+///         2. upgradeAndSetupRevShare() - Combined upgrade + setup (most efficient)
+///         All operations use the default calculator (L1Withdrawer + SuperchainRevenueShareCalculator).
+contract RevShareContractsUpgrader {
+    /// @notice Thrown when portal address is zero
+    error PortalCannotBeZeroAddress();
+
+    /// @notice Thrown when L1Withdrawer recipient is zero address
+    error L1WithdrawerRecipientCannotBeZeroAddress();
+
+    /// @notice Thrown when chain fees recipient is zero address
+    error ChainFeesRecipientCannotBeZeroAddress();
+
+    /// @notice Thrown when gas limit is zero
+    error GasLimitCannotBeZero();
+
+    /// @notice Thrown when array is empty
+    error EmptyArray();
+
+    /// @notice Emitted when a chain's RevShare setup  deposits are completed
+    /// @param portal The portal address for the chain
+    /// @param chainIndex The index of the chain in the configs array
+    event ChainProcessed(address portal, uint256 chainIndex);
+
+    /// @notice Struct for RevShare setup configuration per chain.
+    /// @param portal OptimismPortal2 address for the target L2
+    /// @param l1WithdrawerConfig L1Withdrawer configuration
+    /// @param chainFeesRecipient Chain fees recipient address for the calculator
+    struct RevShareConfig {
+        address portal;
+        FeeSplitterSetup.L1WithdrawerConfig l1WithdrawerConfig;
+        address chainFeesRecipient;
+    }
+
+    /// @notice Upgrades vault and splitter contracts and sets up revenue sharing in one transaction for multiple chains.
+    ///         This is the most efficient path as vaults are initialized with RevShare config from the start.
+    /// @param _configs Array of RevShare configuration structs, one per chain.
+    function upgradeAndSetupRevShare(RevShareConfig[] calldata _configs) external {
+        if (_configs.length == 0) revert EmptyArray();
+
+        for (uint256 i; i < _configs.length; i++) {
+            RevShareConfig calldata config = _configs[i];
+            if (config.portal == address(0)) revert PortalCannotBeZeroAddress();
+            if (config.l1WithdrawerConfig.recipient == address(0)) revert L1WithdrawerRecipientCannotBeZeroAddress();
+            if (config.chainFeesRecipient == address(0)) revert ChainFeesRecipientCannotBeZeroAddress();
+            if (config.l1WithdrawerConfig.gasLimit == 0) revert GasLimitCannotBeZero();
+
+            // Deploy L1Withdrawer and SuperchainRevenueShareCalculator
+            address precalculatedCalculator = FeeSplitterSetup.deployRevSharePeriphery(
+                config.portal, config.l1WithdrawerConfig, config.chainFeesRecipient
+            );
+
+            // Deploy and setup FeeSplitter
+            FeeSplitterSetup.deployAndSetupFeeSplitter(config.portal, precalculatedCalculator);
```

Comment:
A small note on the library approach to be aware of is that you will run into a bit of a verification hell. `RevShareCommon`, `FeeSplitterSetup`, and `FeeVaultUpgrader` will all be external libraries, so you most likely will need to manually link them. At the same time, the `FeeSplitterSetup` and `FeeVaultUpgrader` libraries also use the `RevShareCommon` library. Still, it should be easily manageable.

Likewise, it's a bit of a delegatecall cascade as well, even for functions that perhaps don't need to be delegatecall'd into, like `getSalt` we are delegate calling into. However, the logic should work ok, and the addresses will be warmed up after the first delegatecall, so gas costs should not be significant for further delegatecalls.